### PR TITLE
fix lazy reading of dictionary pages when first column chunk page is skipped

### LINKED ISSUE #192: panic in `SeekToRow` when using dictionary encoding and accessing random row
Some interesting behaviour of SeekToRow captured in the following tests:

1. No dictionary encoding on the columns -> reading random rows works without panics
```golang
func TestSeekToRow_NoDict(t *testing.T) {
	type rowType struct {
		Name utf8string `parquet:","` // no dictionary encoding
	}

	// write samples to in-memory buffer
	buf := new(bytes.Buffer)
	schema := parquet.SchemaOf(new(rowType))
	w := parquet.NewWriter(buf, schema)
	sample := rowType{
		Name: "foo1",
	}
	// write two rows
	w.Write(sample)
	sample.Name = "foo2"
	w.Write(sample)
	w.Close()

	// create reader
	r := parquet.NewReader(bytes.NewReader(buf.Bytes()))

	// read second row
	r.SeekToRow(1)
	row := new(rowType)
	err := r.Read(row)
	if err != nil {
		t.Fatalf("reading row: %v", err)
	}
	// fmt.Println(&sample, row)
	if *row != sample {
		t.Fatalf("read != write")
	}
}
```

2. With dictionary encoding on the columns -> reading rows sequentially works

```golang

func TestSeekToRow_ReadAll(t *testing.T) {
	type rowType struct {
		Name utf8string `parquet:",dict"`
	}

	// write samples to in-memory buffer
	buf := new(bytes.Buffer)
	schema := parquet.SchemaOf(new(rowType))
	w := parquet.NewWriter(buf, schema)
	sample := rowType{
		Name: "foo1",
	}
	// write two rows
	w.Write(sample)
	sample.Name = "foo2"
	w.Write(sample)
	w.Close()

	// create reader
	r := parquet.NewReader(bytes.NewReader(buf.Bytes()))

	// read first row
	r.SeekToRow(0)
	row := new(rowType)
	err := r.Read(row)
	if err != nil {
		t.Fatalf("reading row: %v", err)
	}
	// read second row
	r.SeekToRow(1)
	row = new(rowType)
	err = r.Read(row)
	if err != nil {
		t.Fatalf("reading row: %v", err)
	}
	// fmt.Println(&sample, row)
	if *row != sample {
		t.Fatalf("read != write")
	}
}
```

3. With dictionary encoding on the columns -> reading random rows panics
```golang
func TestSeekToRow_DictReadSecond(t *testing.T) {
	type rowType struct {
		Name utf8string `parquet:",dict"`
	}

	// write samples to in-memory buffer
	buf := new(bytes.Buffer)
	schema := parquet.SchemaOf(new(rowType))
	w := parquet.NewWriter(buf, schema)
	sample := rowType{
		Name: "foo1",
	}
	// write two rows
	w.Write(sample)
	sample.Name = "foo2"
	w.Write(sample)
	w.Close()

	// create reader
	r := parquet.NewReader(bytes.NewReader(buf.Bytes()))

	// read second row
	r.SeekToRow(1)
	row := new(rowType)
	err := r.Read(row)
	if err != nil {
		t.Fatalf("reading row: %v", err)
	}
	// fmt.Println(&sample, row)
	if *row != sample {
		t.Fatalf("read != write")
	}
}
```

```console
Running tool: /usr/local/go/bin/go test -timeout 30s -run ^TestSeekToRow_DictReadSecond$ github.com/segmentio/parquet-go

--- FAIL: TestSeekToRow_DictReadSecond (0.00s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x12dfc49]

goroutine 19 [running]:
testing.tRunner.func1.2({0x13ca760, 0x17d4b70})
	/usr/local/go/src/testing/testing.go:1209 +0x24e
testing.tRunner.func1()
	/usr/local/go/src/testing/testing.go:1212 +0x218
panic({0x13ca760, 0x17d4b70})
	/usr/local/go/src/runtime/panic.go:1038 +0x215
github.com/segmentio/parquet-go.(*indexedPageValues).ReadValues(0xc000244230, {0xc000248000, 0xaa, 0x1})
	/Users/annanay/Desktop/git/go/src/github.com/segmentio/parquet-go/dictionary.go:1039 +0x89
github.com/segmentio/parquet-go.(*columnChunkReader).readValues(0xc0001e3700)
	/Users/annanay/Desktop/git/go/src/github.com/segmentio/parquet-go/column_chunk.go:91 +0xce
github.com/segmentio/parquet-go.readRowsFuncOfLeaf.func1({0xc0001dc738, 0x1, 0x1406c40}, 0x1, {0xc0001e3700, 0x12e3805, 0xc000207e08})
	/Users/annanay/Desktop/git/go/src/github.com/segmentio/parquet-go/column_chunk.go:336 +0x1e7
github.com/segmentio/parquet-go.(*rowGroupRows).ReadRows(0xc0001e36c0, {0xc0001dc738, 0xc00019e5a0, 0x1})
	/Users/annanay/Desktop/git/go/src/github.com/segmentio/parquet-go/row_group.go:336 +0xed
github.com/segmentio/parquet-go.(*reader).ReadRows(0xc0001dc700, {0xc0001dc738, 0x1, 0x1})
	/Users/annanay/Desktop/git/go/src/github.com/segmentio/parquet-go/reader.go:311 +0xb9
github.com/segmentio/parquet-go.(*Reader).Read(0xc0001dc6c0, {0x13a3e80, 0xc000244130})
	/Users/annanay/Desktop/git/go/src/github.com/segmentio/parquet-go/reader.go:192 +0x251
github.com/segmentio/parquet-go_test.TestSeekToRow_DictReadSecond(0xc000187860)
	/Users/annanay/Desktop/git/go/src/github.com/segmentio/parquet-go/reader_test.go:564 +0x229
testing.tRunner(0xc000187860, 0x14527e0)
	/usr/local/go/src/testing/testing.go:1259 +0x102
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:1306 +0x35a
FAIL	github.com/segmentio/parquet-go	0.019s
FAIL

```

(do let me know if you would like me to PR these test cases to this repository!)

### PR DESCRIPTION
**Pull request by achille-roussel**
Fixes #192 

When seeking to a specific row before reading any pages, the dictionary page would not be loaded, and the `Dictionary` value attached to data page would be `nil`, causing a panic when it was accessed later on.

### CODE REVIEW

File: file.go [hunk 1]
```diff
@@ -575,6 +515,115 @@ func (f *filePages) ReadPage() (Page, error) {
...
+	}
+	return f.chunk.column.decodeDataPageV1(DataPageHeaderV1{header.DataPageHeader}, f.dataPage)
+}
+
+func (f *filePages) readDataPageV2(header *format.PageHeader) (Page, error) {
+	if header.DataPageHeaderV2 == nil {
+		return nil, ErrMissingPageHeader
+	}
+	if isDictionaryFormat(header.DataPageHeaderV2.Encoding) && f.dataPage.dictionary == nil {
+		// If the program seeked to a row passed the first page,
```
Comment (kevinburkesegment) [inline file.go:576]: technically this verb is "sought" but that's also awkward hahaha

File: file.go [hunk 2]
```diff
@@ -575,6 +515,115 @@ func (f *filePages) ReadPage() (Page, error) {
...
+		// the dictionary page may not have been seen, in which
+		// case we have to lazily load it despite
```
Comment (kevinburkesegment) [inline file.go:578]: sentence fragment here

**Comment by kevinburkesegment** [comment]
Should we keep a CONTRIBUTORS list with people who have contributed a change or reported an error?

**Comment by achille-roussel** [comment]
> Should we keep a CONTRIBUTORS list with people who have contributed a change or reported an error?

Github generates this already https://github.com/segmentio/parquet-go/graphs/contributors, do you think it's enough or is there value in maintaining a separate file?

**Pull request merged**